tick/optim/prox/src/prox_sorted_l1.cpp

```diff
@@ -8,173 +8,149 @@
-    // Sort sub_coeffs with decreasing absolute values, keeping the sorting index
-    ArrayDouble sub_coeffs_sorted = sort_abs(sub_coeffs, idx, false);
-
-    ArrayDouble sub_coeffs_abs_sort(size);
-    // Multiply weights by the threshold, compute abs and sign of sub_coeffs,
-    // and compute abs(sub_coeffs_sorted)
-    for (ulong i = 0; i < size; i++) {
-        weights_copy[i] *= thresh;
-        sub_coeffs_abs_sort[i] = std::abs(sub_coeffs_sorted[i]);
-        double sub_coeffs_i = sub_coeffs[i];
-        if (sub_coeffs_i >= 0) {
-            sub_coeffs_sign[i] = 1;
-            sub_coeffs_abs[i] = sub_coeffs_i;
-        } else {
-            sub_coeffs_sign[i] = -1;
-            sub_coeffs_abs[i] = -sub_coeffs_i;
-        }
-    }
-    // Where do the crossing occurs?
-    ulong crossing = 0;
-    for (ulong i = 0; i < size; i++) {
-        if (sub_coeffs_abs_sort[i] > weights_copy[i]) {
-            crossing = i;
-        }
-    }
-    ulong n_sub_coeffs;
-    if (crossing > 0) {
-        n_sub_coeffs = crossing + 1;
+  // If necessary, compute weights
+  compute_weights();
+  ulong size = end - start;
+  double thresh = t;
+
+  ArrayDouble sub_coeffs = view(coeffs, start, end);
+  ArrayDouble sub_out = view(out, start, end);
+  sub_out.fill(0);
+
+  ArrayDouble weights_copy(weights);
+
+  ArrayShort sub_coeffs_sign(size);
+  ArrayDouble sub_coeffs_abs(size);
+  // Will contain the indexes that sort abs(sub_coeffs) in decreasing order
+  ArrayULong idx(size);
+  // Sort sub_coeffs with decreasing absolute values, keeping the sorting index
+  ArrayDouble sub_coeffs_sorted = sort_abs(sub_coeffs, idx, false);
+
+  ArrayDouble sub_coeffs_abs_sort(size);
+  // Multiply weights by the threshold, compute abs and sign of sub_coeffs,
+  // and compute abs(sub_coeffs_sorted)
+  for (ulong i = 0; i < size; i++) {
+    weights_copy[i] *= thresh;
+    sub_coeffs_abs_sort[i] = std::abs(sub_coeffs_sorted[i]);
+    double sub_coeffs_i = sub_coeffs[i];
+    if (sub_coeffs_i >= 0) {
+      sub_coeffs_sign[i] = 1;
+      sub_coeffs_abs[i] = sub_coeffs_i;
     } else {
-        n_sub_coeffs = size;
-    }
-
-    ArrayDouble subsub_coeffs = view(sub_coeffs_abs_sort, 0, n_sub_coeffs);
-    ArrayDouble subsub_out(n_sub_coeffs);
-    subsub_out.fill(0);
-
-    prox_sorted_l1(subsub_coeffs, weights_copy, subsub_out);
-
-    for (ulong i = 0; i < n_sub_coeffs; i++) {
-        sub_out[idx[i]] = subsub_out[i];
+      sub_coeffs_sign[i] = -1;
+      sub_coeffs_abs[i] = -sub_coeffs_i;
     }
-    for (ulong i = 0; i < size; i++) {
-        sub_out[i] *= sub_coeffs_sign[i];
+  }
+  // Where do the crossing occurs?
+  ulong crossing = 0;
+  for (ulong i = 0; i < size; i++) {
+    if (sub_coeffs_abs_sort[i] > weights_copy[i]) {
+      crossing = i;
     }
+  }
+  ulong n_sub_coeffs;
+  if (crossing > 0) {
+    n_sub_coeffs = crossing + 1;
+  } else {
+    n_sub_coeffs = size;
+  }
+
+  ArrayDouble subsub_coeffs = view(sub_coeffs_abs_sort, 0, n_sub_coeffs);
```

Comment:
What happens if `crossing == size`? Then `n_sub_coeffs == size + 1 `and it looks like we could go out of bounds.